Mono musl support

ayakael opened the pull request:
## Descriptions
Attempt at musl support for mono as to allow s390x on Alpine Linux.

There seems to be issues with mono on musl. Outside perspective would be much appreciated.

## Related issues:
`System.InsufficientExecutionStackException` when building many packages (https://github.com/dotnet/roslyn/issues/64423)

Fixes https://github.com/dotnet/runtime/issues/76805

am11 commented on src/mono/mono.proj:
```diff
@@ -554,18 +560,19 @@
-      <_LinuxAbi Condition="'$(TargetsAndroid)' != 'true' and '$(TargetsLinuxBionic)' != 'true'">gnu</_LinuxAbi>
-      <_LinuxAbi Condition="'$(TargetsAndroid)' == 'true' or '$(TargetsLinuxBionic)' == 'true'">android</_LinuxAbi>
+      <_LinuxAbi Condition="'$(TargetsAndroid)' != 'true' and '$(TargetsLinuxBionic)' != 'true' and '$(TargetsLinuxMusl)' != 'true'">linux-gnu</_LinuxAbi>
+      <_LinuxAbi Condition="'$(TargetsAndroid)' != 'true' and '$(TargetsLinuxBionic)' != 'true' and '$(TargetsLinuxMusl)' == 'true'">alpine-linux-musl</_LinuxAbi>
```
Is there a `{arch}-linux-musl-{tool}` variant available (without the `alpine-` part), which would work for AlpineLine and VoidLinuxMusl alike?

ayakael replied:
Hmm good question. I don't know how void does their crosscompilation, as really this code is only relevant then.

am11 replied:
I use the portable linux-x64 and linux-musl-x64 installer binaries on Void Linux amd64 and musl amd64 respectively (since https://github.com/dotnet/runtime/issues/28265), it works well. I haven't cross-compiled mono for Void. :)

Judging by the `system types` at https://gcc.gnu.org/install/configure.html, distro can choose to use `linux-musl` target.

ayakael replied:
Thus if TargetsLinuxMusl _LinuxAbi should be linux-musl. Thus perhaps I should add in a TargetsLinuxAlpine check.

ayakael replied:
Indeed, Alpine's crosscompilation framework uses `{arch}-alpine-linux-musl-{tool}`. If we were to want to target `void-{armv6,s390x}`, then I'd think that having seperate checks for `alpine`and `linux-musl` would be relevant. Given the former being unlikely anytime soon, and judging by having already included alpine-specific code in `toolchain.cmake`, I'd consider it out-of-scope of this PR.

akoeplinger replied:
I think we should be forward looking and add an explicit `TargetsAlpineLinux` property and check (it's an easy addition to the root Directory.Build.props).

ayakael commented on src/mono/mono.proj:
```diff
@@ -470,6 +470,12 @@
+      <_MonoAOTCXXFLAGS Include="-lucontext" />
+      <_MonoCFLAGS Include="-lucontext" />
+      <_MonoCXXFLAGS Include="-lucontext" />
+    </ItemGroup>
```
The following is due to `ucontext` symbols being provided by a seperate `libucontext` package, as `musl` does not natively support it.

am11 commented on src/mono/mono/utils/mono-context.h:
```diff
@@ -11,6 +11,14 @@
+ * Handle non-gnu libc versions with nothing in features.h 
+ * We have no idea what they're compatible with, so always fail.
+ */
+#ifndef __GLIBC_PREREQ
```
Probably better to just change its (single) usage below:

```diff
- #if __GLIBC_PREREQ(2, 26)
+ #if defined(__GLIBC_PREREQ) && __GLIBC_PREREQ(2, 26)
```
or `!defined(__GLIBC_PREREQ) ||` depending on the availability of `struct ucontext`.

ayakael replied:
Pushed

ayakael commented:
The following bugs exist for mono on musl:
* https://github.com/dotnet/runtime/issues/76523
* https://github.com/dotnet/runtime/issues/76543

A fix for one of them likely involves setting stack size properly due to musl's really low default of 128K rather than the usual 8M.

ayakael commented:
With progress done on https://github.com/dotnet/runtime/issues/76805, I've got a full build of s390x working on dotnet6. They are still some issues on dotnet7, which indeed only occurs on `s390x`:
```
The process cannot access the file '/builds/ayakael/aports/testing/dotnet7-stage0/src/dotnet-d41bfecf5090e9163aa2da251246abed9e756e53/src/runtime/artifacts/obj/System.IO.Ports' because it is being used by another process
```

vargaz commented on src/mono/mono/metadata/domain.c:
```diff
@@ -128,6 +136,8 @@ mono_init_internal (const char *root_domain_name)
 	mono_counters_register ("Max HashTable Chain Length", MONO_COUNTER_INT|MONO_COUNTER_METADATA, &mono_g_hash_table_max_chain_length);
 
+	ensure_stack_size ();
+
```
Why is this needed ? It doesn't look like something that would work on all platforms.

ayakael replied:
This fixes https://github.com/dotnet/runtime/issues/76805, but indeed this is something that, like in coreclr, should be turned on with a `ENSURE_PRIMARY_STACK_SIZE` definition.

ayakael replied:
Pushed conditional use of `ensure_stack_size`, modeled after coreclr.

vargaz replied:
I'd suggest moving this to mini_init () in mini-runtime.c, thats where all the arch specific execution engine stuff is.

ayakael committed:
Fixes for building mono on musl-based systems

ayakael commented:
This is now again ready for review. I fixed the alpine detection logics, and added in some changes to `toolchain.cmake` as to allow crossbuilding to all architectures already included. Since alpine doesn't support multi-arch libraries, `-m32` mode is disabled when crossbuilding on arch.

One remaining issue is that `lld` is something used even when it doesn't work (i.e, x86 lld is broken on our build, and lld for s390x isn't supported on Alpine). It'd be nice if there could be a flag to disable ldd usage. A workaround I've used is the following:
```diff
From b99f2106370e1af3ecd116576341b94319cbd332 Mon Sep 17 00:00:00 2001
From: "build@apk-groulx" <build@apk-groulx.praxis>
Date: Sun, 18 Sep 2022 00:45:20 +0000
Subject: [PATCH 1/1] s390x fixes

---
 src/runtime/eng/common/cross/toolchain.cmake  |  7 ++++++-
 .../eng/common/native/init-compiler.sh        |  2 +-
 6 files changed, 26 insertions(+), 14 deletions(-)

diff --git a/src/runtime/eng/common/native/init-compiler.sh b/src/runtime/eng/common/native/init-compiler.sh
index 41a26d802..f81f3257f 100644
--- a/src/runtime/eng/common/native/init-compiler.sh
+++ b/src/runtime/eng/common/native/init-compiler.sh
@@ -135,7 +135,7 @@ fi
 # Only lld version >= 9 can be considered stable
 if [[ "$compiler" == "clang" && "$majorVersion" -ge 9 ]]; then
     if "$CC" -fuse-ld=lld -Wl,--version >/dev/null 2>&1; then
-        LDFLAGS="-fuse-ld=lld"
+        [ -f "$ROOTFS_DIR/usr/bin/lld" ] && LDFLAGS="-fuse-ld=lld"
     fi
 fi

-- 
2.37.1
```

Is there such a flag?

ayakael commented:
Once changes look good for changes in `eng/common`, I'll take them out of this PR and send them to `Arcade`

am11 commented:
> x86 lld is broken on our build

Broken toolchain should be fixed instead of adding workarounds in the code. You can uninstall lld from the problematic machine such that `"$CC" -fuse-ld=lld -Wl,--version >/dev/null` does not succeed. `init-compiler` has no knowledge of rootfs.

am11 commented on eng/common/cross/toolchain.cmake:
```diff
@@ -25,8 +25,6 @@ if(TARGET_ARCH_NAME STREQUAL "arm")
   if(EXISTS ${CROSS_ROOTFS}/usr/lib/gcc/armv7-alpine-linux-musleabihf)
     set(TOOLCHAIN "armv7-alpine-linux-musleabihf")
-  elseif(EXISTS ${CROSS_ROOTFS}/usr/lib/gcc/armv6-alpine-linux-musleabihf)
-    set(TOOLCHAIN "armv6-alpine-linux-musleabihf")
```
Why is this removed?

ayakael replied:
My mistake, this wasn't supposed to be removed.

ayakael commented:
> > x86 lld is broken on our build
> 
> Broken toolchain should be fixed instead of adding workarounds in the code. You can uninstall lld from the problematic machine such that `"$CC" -fuse-ld=lld -Wl,--version >/dev/null` does not succeed. `init-compiler` has no knowledge of rootfs.

Copy that. I've found a fix for the `x86` stuff (setting add_linker_flag(-Wl,-z,notext) ). Exploring s390x.

am11 commented on src/mono/mono/mini/mini-runtime.c:
```diff
@@ -4408,6 +4408,31 @@ mini_llvm_init (void)
+  
+ --*/
+static MONO_NO_OPTIMIZATION MONO_NEVER_INLINE void
+ensure_stack_size (void)
+{
+	const int default_size = 5 * 1024 * 1024;
+	volatile uint8_t *s = (uint8_t *)g_alloca(default_size);
```
nit: in coreclr pal (and nativeaot), we pass the size as argument from call site. It would be nice to keep it aligned:

```c
ensure_stack_size (size_t size)
{
	volatile uint8_t *s = (uint8_t *)g_alloca(size);
```

and call it with `ensure_stack_size(1536 * 1024)`.

ayakael replied:
Pushed the requested change, thanks :)

vargaz commented on src/mono/mono/mini/CMakeLists.txt:
```diff
@@ -554,3 +554,9 @@ if(NOT DISABLE_EXECUTABLES)
 endif()
+
+if(HOST_ALPINE_LINUX)
+  # On Alpine Linux, we need to ensure that the reported stack range for the primary thread is
```
This should probably be moved to the main CMakeFiles.txt where the other if(HOST) definitions are.

vargaz commented:
The mono/ changes look ok to me.

akoeplinger commented on eng/native/configurecompiler.cmake:
```diff
@@ -448,6 +452,7 @@ if (CLR_CMAKE_HOST_UNIX)
     add_compile_options(-Wno-incompatible-ms-struct)
 
     add_compile_options(-Wno-reserved-identifier)
+    add_compile_options(-Wno-sign-conversion)
```
why do you need to disable this warning?

ayakael replied:
Building for `ppc64le` with clang-15 gives out `Wsign-conversion` errors. I can reproduce the error if you require the exact error message.

am11 replied:
Did you observe those warnings outside of `src/mono`? If not, please move this suppression to one of the `CMakeLists.txt` file under src/mono. Note that we are building coreclr with clang-15 in arm64 CI leg.

ayakael replied:
Here is the exact error:
```
/var/build/dotnet7/testing/dotnet7-stage0/src/dotnet-e6dd91c290b808f971a1ac69c2fb29395bbf1051/src/runtime/src/native/libs/System.Native/pal_io.c(1270,36): error GBBD8A67C: implicit conversion changes signedness: 'unsigned long' to 'int' [-Wsign-conversion] [/var/build/dotnet7/testing/dotnet7-stage0/src/dotnet-e6dd91c290b808f971a1ac69c2fb29395bbf1051/src/runtime/src/native/libs/build-native.proj]
/var/build/dotnet7/testing/dotnet7-stage0/src/dotnet-e6dd91c290b808f971a1ac69c2fb29395bbf1051/src/runtime/src/native/libs/System.Native/pal_io.c(1270,36): error GBBD8A67C: implicit conversion changes signedness: 'unsigned long' to 'int' [-Wsign-conversion] [/var/build/dotnet7/testing/dotnet7-stage0/src/dotnet-e6dd91c290b808f971a1ac69c2fb29395bbf1051/src/runtime/src/native/libs/build-native.proj]
```

am11 replied:
`(int)FICLONE` at:
https://github.com/dotnet/runtime/blob/37191cd6bcd0d7062b1c5b84fd857f8e261c25e9/src/native/libs/System.Native/pal_io.c#L1270

ayakael replied:
That did it, thanks! Pushed without no-sign-conversion.

akoeplinger commented on eng/native/configurecompiler.cmake:
```diff
@@ -193,6 +193,10 @@ if(CLR_CMAKE_HOST_LINUX)
   add_linker_flag(-Wl,-z,relro,-z,now)
+  if(CLR_CMAKE_HOST_ALPINE_LINUX)
+    add_linker_flag(-lucontext)
+	add_linker_flag(-Wl,-z,notext)
```
can you please add a comment explaining why we need this on alpine?

ayakael replied:
Absolutely.

akoeplinger commented on src/mono/mono/mini/mini-runtime.c:
```diff
@@ -4565,6 +4589,11 @@ mini_init (const char *filename)
 #endif
 
+#ifdef ENSURE_PRIMARY_STACK_SIZE 
+	// TODO: https://github.com/dotnet/runtimelab/issues/791
```
looks like this issue was moved:

```suggestion
	// TODO: https://github.com/dotnet/runtime/issues/72920
```

akoeplinger commented on src/mono/mono.proj:
```diff
@@ -464,11 +464,21 @@
     <!-- Linux options -->
     <ItemGroup Condition="'$(TargetsLinux)' == true">
       <_MonoCFLAGS Include="-Wl,--build-id=sha1" />
+      <_MonoCFLAGS Include="-Wno-int-conversion" />
```
same question as above, why does this need to be disabled?

ayakael replied:
Building for `ppc64le` with clang-15 gives out `Wint-conversion` errors. I can reproduce the error if you require the exact error message.

ayakael replied:
Error related to `Wint-conversion` is the following:
```
  [ 97%] Building C object mono/mini/CMakeFiles/monosgen-objects.dir/var/build/dotnet7/testing/dotnet7-stage0/src/dotnet-e6dd91c290b808f971a1ac69c2fb29395bbf1051/src/runtime/artifacts/obj/_version.c.o
  /var/build/dotnet7/testing/dotnet7-stage0/src/dotnet-e6dd91c290b808f971a1ac69c2fb29395bbf1051/src/runtime/src/mono/mono/mini/tramp-ppc.c:675:11: error: incompatible integer to pointer conversion initializing 'guint8 *' (aka 'unsigned char *') with an expression of type 'unsigned long' [-Wint-conversion]
                  guint8 *thunk = GET_MEMORY_SLOT_THUNK_ADDRESS((guint32*)(code - 32));
                          ^       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  /var/build/dotnet7/testing/dotnet7-stage0/src/dotnet-e6dd91c290b808f971a1ac69c2fb29395bbf1051/src/runtime/src/mono/mono/mini/tramp-ppc.c:678:11: error: incompatible integer to pointer conversion initializing 'guint8 *' (aka 'unsigned char *') with an expression of type 'unsigned long' [-Wint-conversion]
                  guint8 *thunk = GET_MEMORY_SLOT_THUNK_ADDRESS((guint32*)(code - 4));
                          ^       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  2 errors generated.
```

ayakael replied:
Related to this piece of code: https://github.com/dotnet/runtime/blob/42bebc934bc86d4b612190810cfbbcb0cb9f7d3d/src/mono/mono/mini/tramp-ppc.c#L673-L680 added by @akoeplinger

uweigand replied:
The `GET_MEMORY_SLOT_THUNK_ADDRESS` macro returns a `guint64`, so I guess this simply needs a cast to `guint8 *`.  @alhad-deshpande can you look into this?

ayakael replied:
Looks like this was fixed in https://github.com/dotnet/runtime/pull/78102

am11 commented on eng/native/configurecompiler.cmake:
```diff
@@ -193,6 +193,10 @@ if(CLR_CMAKE_HOST_LINUX)
   add_linker_flag(-Wl,-z,relro,-z,now)
+  if(CLR_CMAKE_HOST_ALPINE_LINUX)
+    add_linker_flag(-lucontext)
+	add_linker_flag(-Wl,-z,notext)
```
What doesn't work without libucontext? This file is used by coreclr, libs and corehost as well, none of which every required this additional dependency.

libucontext is a separate package dependency which we should try not to introduce unless it's necessary (in which case it should be moved near to the cmake file of component needing it).

ayakael replied:
If memory serves, libunwind was needing it on `x86`.

ayakael replied:
I was wrong, ildasm also wants to use `setcontext` symbols, which are included natively in `glibc`
```
 [100%] Linking CXX executable ilasm
  ld.lld: error: undefined symbol: setcontext
  >>> referenced by Gos-linux.c:309 (/var/build/dotnet7/testing/dotnet7-stage0/src/dotnet-e6dd91c290b808f971a1ac69c2fb29395bbf1051/src/runtime/src/native/external/libunwind/src/x86/Gos-linux.c:309)
  >>>               Los-linux.c.o:(_ULx86_local_resume) in archive ../pal/src/libcoreclrpal.a
  clang-15: error: linker command failed with exit code 1 (use -v to see invocation)
  make[3]: *** [ilasm/CMakeFiles/ilasm.dir/build.make:267: ilasm/ilasm] Error 1
  make[2]: *** [CMakeFiles/Makefile2:2715: ilasm/CMakeFiles/ilasm.dir/all] Error 2
  make[2]: *** Waiting for unfinished jobs....
  [100%] Linking CXX executable ildasm
  ld.lld: error: undefined symbol: setcontext
  >>> referenced by Gos-linux.c:309 (/var/build/dotnet7/testing/dotnet7-stage0/src/dotnet-e6dd91c290b808f971a1ac69c2fb29395bbf1051/src/runtime/src/native/external/libunwind/src/x86/Gos-linux.c:309)
  >>>               Los-linux.c.o:(_ULx86_local_resume) in archive ../../pal/src/libcoreclrpal.a
  clang-15: error: linker command failed with exit code 1 (use -v to see invocation)
  make[3]: *** [ildasm/exe/CMakeFiles/ildasm.dir/build.make:385: ildasm/exe/ildasm] Error 1
  make[2]: *** [CMakeFiles/Makefile2:2750: ildasm/exe/CMakeFiles/ildasm.dir/all] Error 2
  make[1]: *** [CMakeFiles/Makefile2:1621: CMakeFiles/iltools.dir/rule] Error 2
  make: *** [Makefile:234: iltools] Error 2
  ~/dotnet7/testing/dotnet7-stage0/src/dotnet-e6dd91c290b808f971a1ac69c2fb29395bbf1051/src/runtime/src/coreclr
  Failed to build "CoreCLR component".
```

ayakael replied:
Too quick to the punch, indeed still libunwind but the change in location isn't working for some reason.

am11 commented on src/coreclr/pal/src/CMakeLists.txt:
```diff
@@ -318,6 +318,11 @@ if(CLR_CMAKE_TARGET_LINUX)
   endif(CLR_CMAKE_USE_SYSTEM_LIBUNWIND)
 
+  # bundled libunwind requires using libucontext on alpine and x86
+  if(CLR_CMAKE_HOST_ALPINE_LINUX AND CLR_CMAKE_HOST_ARCH_I386)
+    add_linker_flag(-lucontext)
+  endif(CLR_CMAKE_HOST_ALPINE_LINUX AND CLR_CMAKE_HOST_ARCH_I386)
```
We are using `TARGET` in this scope for cross compilation:
```suggestion
  if(CLR_CMAKE_TARGET_ALPINE_LINUX AND CLR_CMAKE_TARGET_ARCH_I386)
    target_link_libraries(coreclrpal ucontext)
  endif(CLR_CMAKE_TARGET_ALPINE_LINUX AND CLR_CMAKE_TARGET_ARCH_I386)
```

(note that coreclr, libs and corehost have different host / target semantics than mono; mono uses autoconf semantics {host,build,target} as opposed to {host,target}, which is why we see HOST under mono most of the times and TARGET in rest of the repo)

ayakael commented:
@am11 Do the arcade-level changes look good to you? If so, I'll take them out and send-em upstream.

am11 commented:
@ayakael, if everything works on your end without changing `eng/common/cross/toolchain.cmake`, then I recommend to revert that file.

The reason being; toolchain.cmake and build-rootfs.sh are tied together, i.e. we use a specific [cross-compilation cmake setup](https://cmake.org/cmake/help/latest/manual/cmake-toolchains.7.html#toolchain-features) which expects certain files to be present at specific location under the directory pointed by `ROOTFS_DIR` environment variable. Therefore, changing one but not the other indicates that we are not testing it atm. Until that happens, you can ignore `toolchain.cmake` and `tryrun.cmake`.

ayakael commented:
> @ayakael, if everything works on your end without changing `eng/common/cross/toolchain.cmake`, then I recommend to revert that file.
> 
> The reason being; toolchain.cmake and build-rootfs.sh are tied together, i.e. we use a specific [cross-compilation cmake setup](https://cmake.org/cmake/help/latest/manual/cmake-toolchains.7.html#toolchain-features) which expects certain files to be present at specific location under the directory pointed by `ROOTFS_DIR` environment variable. Therefore, changing one but not the other indicates that we are not testing it atm. Until that happens, you can ignore `toolchain.cmake` and `tryrun.cmake`.

Understood. The `toolchain.cmake`and `tryrun.cmake` fixes crossbuilding on Alpine, but granted I havn't tested `build-rootfs.sh` with those changes.

ayakael commented:
All changes under `eng/common` have been reverted, and current state of MR has been tested on Alpine's end and its looking good for us.

am11 commented:
> The `toolchain.cmake`and `tryrun.cmake` fixes crossbuilding on Alpine, but granted I havn't tested `build-rootfs.sh` with those changes.

Thanks. We can always add them later after testing it locally with `ROOTFS_DIR=<PATH_TO_ARCH_DIR> ./build.sh -cross -arch <ARCH>`.

ayakael commented:
> > The `toolchain.cmake`and `tryrun.cmake` fixes crossbuilding on Alpine, but granted I havn't tested `build-rootfs.sh` with those changes.
> 
> Thanks. We can always add them later after testing it locally with `ROOTFS_DIR=<PATH_TO_ARCH_DIR> ./build.sh -cross -arch <ARCH>`.

For what it's worth we do exactly that on Alpine. We run:
```
ROOTFS_DIR="$CBUILDROOT" ./build.sh \
		-c Release \
		-clang \
		-arch $_dotnet_target \
		-cross \
		/p:VersionSuffix=${_runtimever##*-} \
		/p:NoPgoOptimize=true \
		/p:EnableNgenOptimization=false \
		/p:GitCommitHash=$(cat ./.git/HEAD) \
```
I suppose we would have to make sure it runs well on `linux-x64`

ayakael commented:
The only remaining issue that I see is investigating `Wno-int-conversion`. I'm pretty sure it's a matter of a quick change, but due to my lack of confidence in regards to C, I'm not well placed to propose a change for that. I pulled the relevant code up-top.

am11 commented on src/mono/mono.proj:
```diff
@@ -464,11 +464,21 @@
       <_MonoAOTCXXFLAGS Include="-Wl,--build-id=sha1" />
+      <_MonoAOTCXXFLAGS Include="-Wno-int-conversion" />
+    </ItemGroup>
+    <ItemGroup Condition="'$(TargetsLinuxMusl)' == 'true'">
+      <_MonoAOTCFLAGS Include="-lucontext" />
+      <_MonoAOTCXXFLAGS Include="-lucontext" />
+      <_MonoCFLAGS Include="-lucontext" />
+      <_MonoCXXFLAGS Include="-lucontext" />
     </ItemGroup>
```
```suggestion
```

not needed, right?

ayakael replied:
These were added as without it the libucontext library isn't pulled automatically on musl.

am11 replied:
You mentioned that (HP) libunwind was requiring libucontext on i386 and PowerPC64, and added that condition in `src/coreclr/pal/src/CMakeLists.txt`. Was that not enough, or is there something else requiring libucontext in mono? (for all architectures?)

ayakael replied:
There's a mono specific requirement for `ucontext` on `s390x`:
```
  /usr/bin/s390x-alpine-linux-musl-ld: libmonosgen-2.0.a(exceptions-s390x.c.o): in function `mono_arch_get_restore_context':
  exceptions-s390x.c:(.text+0x14de): undefined reference to `setcontext'
  clang-15: error: linker command failed with exit code 1 (use -v to see invocation)
  make[2]: *** [mono/mini/CMakeFiles/mono-sgen.dir/build.make:194: mono/mini/mono-sgen] Error 1
  make[1]: *** [CMakeFiles/Makefile2:660: mono/mini/CMakeFiles/mono-sgen.dir/all] Error 2
  make: *** [Makefile:136: all] Error 2
/var/build/dotnet7/testing/dotnet7-stage0/src/dotnet-v7.0.100-rtm.22521.12/src/runtime/src/mono/mono.proj(551,5): error MSB3073: The command "TARGET_BUILD_ARCH=s390x PKG_CONFIG_PATH=/var/build/sysroot-s390x//usr/lib/s390x-linux-gnu/pkgconfig cmake --build . --target install --config Release --parallel 36" exited with code 2.
```
Can probably remove the CXXFLAGS component given that being a C file.

am11 replied:
Ok, `setcontext` is only used on s390x in mono, so if there is no requirement for other architectures, we can append this condition: `and '$(TargetArchitecture)' == 's390x'"`.

ayakael replied:
Done

am11 commented on src/mono/mono.proj:
```diff
@@ -464,11 +464,21 @@
       <_MonoAOTCFLAGS Include="-Wl,--build-id=sha1" />
+      <_MonoAOTCFLAGS Include="-Wno-int-conversion" />
       <_MonoAOTCXXFLAGS Include="-Wl,--build-id=sha1" />
+      <_MonoAOTCXXFLAGS Include="-Wno-int-conversion" />
```
Are these CXXFLAGS needed as well? Mono is mainly a C project with a few C++ files compiled when `mono.mscordbi`  subset is built (it is optional and a small component). Mirroring every C flag to C++ is not needed.

ayakael replied:
Ideally, I'm looking to remove the necessity for `-Wno-int-conversion`. I'll try removing the CXX flags and see how that goes.

am11 commented on src/mono/mono/utils/mono-context.h:
```diff
@@ -11,6 +11,14 @@
 #ifndef __MONO_MONO_CONTEXT_H__
 #define __MONO_MONO_CONTEXT_H__
 
+/* 
+ * Handle non-gnu libc versions with nothing in features.h 
+ * We have no idea what they're compatible with, so always fail.
+ */
+#ifndef __GLIBC_PREREQ
+# define __GLIBC_PREREQ(x,y) 0
+#endif
```
At line 795:

```diff
- #if __GLIBC_PREREQ(2, 26)
+ #if defined(__GLIBC_PREREQ) && __GLIBC_PREREQ(2, 26)
```
and delete this block. This macro is defined by glibc and we have only one usage; we don't need to define it (just to return false).

ayakael replied:
Previous attempts at this proposal had yielded negative results:
```
  /var/build/dotnet7/testing/dotnet7-stage0/src/dotnet-v7.0.100-rtm.22521.12/src/runtime/src/mono/mono/mini/../../mono/utils/mono-context.h:787:32: error: function-like macro '__GLIBC_PREREQ' is not defined
  #if defined(__GLIBC_PREREQ) && __GLIBC_PREREQ(2, 26)
```

am11 replied:
```c
#if defined (__GLIBC_PREREQ)
#if __GLIBC_PREREQ(2, 26)
...
#endif // __GLIBC_PREREQ(2, 26)
#endif // defined (__GLIBC_PREREQ)
```

will fix the error.

ayakael replied:
Since __GLIBC_PREREQ is not defined, it never gets to go to else definition of ucontext, thus necessary to do this:
```c
#if defined(__GLIBC_PREREQ)
#if __GLIBC_PREREQ(2, 26)
typedef ucontext_t MonoContext;
#else
typedef struct ucontext MonoContext;
#endif // __GLIBC_PREREQ(2, 26)
#else
typedef struct ucontext MonoContext;
#endif // defined (__GLIBC_PREREQ)
```
Not very elegant, which is why previous fix just defined __GLIBC_PREREQ, but assuring it will always fail.

am11 replied:
Ok, sounds good.

am11 reviewed (approved):
LGTM, thank you!

tmds commented on src/mono/mono/mini/mini-runtime.c:
```diff
@@ -4383,6 +4383,30 @@ mini_llvm_init (void)
+  
+ Abstract: 
+   This fixes a problem on MUSL where the initial stack size reported by the 
+   pthread_attr_getstack is about 128kB, but this limit is not fixed and 
```
When does the stack size get reported as 128kB?

If I run the following code in an Alpine container, it reports the size to be 8MB (8388608), as set when the thread was created.

```
#include <pthread.h>
#include <stdio.h>

void* thread_fn(void*)
{
    printf("hello from thread\n");
    return 0;
}

int main()
{
    pthread_attr_t attr;
    pthread_t thread;
    int res;
    size_t set_stack_size = 8*1024*1024;
    size_t stack_size;

    res = pthread_attr_init (&attr);

    res = pthread_attr_setstacksize (&attr, set_stack_size);

    res = pthread_create (&thread, &attr, thread_fn, NULL);

    res = pthread_attr_getstacksize (&attr, &stack_size);

    pthread_join(thread, NULL);

    printf("%zu\n", stack_size);

    return 0;
}

```

tmds replied:
nvmd, here it is:
```c
#define _GNU_SOURCE   
#include <pthread.h>
#include <stdio.h>

int main()
{
    pthread_attr_t attr;
    pthread_t thread = pthread_self();

    res = pthread_attr_init (&attr);

    res = pthread_getattr_np (thread, &attr);

    res = pthread_attr_getstacksize (&attr, &stack_size);

    printf("%zu\n", stack_size);

    return 0;
}
```

```
$ ./a.out 
131072
```

akoeplinger commented on src/mono/CMakeLists.txt:
```diff
@@ -216,6 +216,33 @@ elseif(CMAKE_SYSTEM_NAME STREQUAL "Linux")
+  if(EXISTS ${LINUX_ID_FILE})
+      execute_process(
+          COMMAND bash -c "source ${LINUX_ID_FILE} && echo \$ID"
+          OUTPUT_VARIABLE LINUX_ID
```
please call this CLR_CMAKE_LINUX_ID, we eventually want to unify Mono's CMake files with the stuff in eng/native so it'd be good to keep the variable names.

akoeplinger replied:
and maybe also add a TODO comment to unify this

akoeplinger commented on src/mono/mono.proj:
```diff
@@ -470,6 +470,10 @@
       <_MonoAOTCFLAGS Include="-Wl,--build-id=sha1" />
       <_MonoAOTCXXFLAGS Include="-Wl,--build-id=sha1" />
     </ItemGroup>
+    <ItemGroup Condition="'$(TargetsLinuxMusl)' == 'true' and '$(TargetArchitecture)' == 's390x'">
```
ideally we'd do this in CMake instead of passing it in here.

akoeplinger replied:
looking at the change you made to the coreclr pal you only did this for x86/powerpc64 there?

akoeplinger replied:
interesting, thanks for the explanation

akoeplinger reviewed (commented):
This is looking pretty good, just a few comments

ayakael commented:
@akoeplinger Thanks! I addressed your comments :)

am11 commented on Directory.Build.props:
```diff
@@ -207,6 +207,9 @@
     <TargetsLinuxBionic Condition="$(OutputRid.StartsWith('linux-bionic'))">true</TargetsLinuxBionic>
+    <TargetsLinuxMusl Condition="$(OutputRid.StartsWith('linux-musl')) or $(OutputRid.StartsWith('alpine'))">true</TargetsLinuxMusl>
+    <!-- TODO: Make distinction between AlpineLinux and LinuxMusl for Void -->
+    <TargetsAlpineLinux Condition="$(OutputRid.StartsWith('linux-musl')) or $(OutputRid.StartsWith('alpine'))">true</TargetsAlpineLinux>
```
`TargetsLinuxMusl` was enough I think; aligns with `TargetsLinuxBionic`. It covers both portable and non-portable use-cases which we care about.

akoeplinger replied:
this is what I had in mind:

```suggestion
    <TargetsAlpineLinux Condition="$(OutputRid.StartsWith('alpine'))">true</TargetsAlpineLinux>
```

we'd only use it in cases where something applies only to alpine, not all musl platforms.

ayakael replied:
This doesn't work as building on Alpine as portable build yields `linux-musl-$arch` for `OutputRid`. We thus need to check for non-portable RID, but AFAIK there is no variable that'll match `alpine.3.17-$arch` when building a portable build of runtime. It'll eventually be necessary to differentiate as on Void `_LinuxAbi` should not be `alpine-linux-musl`. As Void hasn't yet packaged dotnet, this is a change that is out-of-scope of this PR. I propose keeping as-is, and eventually come back to this.

ayakael replied:
Copy that, applying your suggestion of going back to TargetsLinuxMusl.

ayakael replied:
@akoeplinger Are you okay with these changes? We can double back once Void wants to target mono-based platforms. By then, I expect we'll have merged with `eng/native` anyways.

ayakael replied:
Discovered this today, `_portableOS` is not always `linux-musl` on Alpine. When building runtime in source-build, `RuntimeOS` is set as `alpine.3.17` which then keeps `_portableOS` as `linux` as `RuntimeOS` isn't equal to `linux-musl`. I've thus reverted to the original `TargersLinuxMusl` logics.

am11 replied:
> which then keeps `_portableOS` as `linux` 

Then it is a bug and we should handle it above. https://github.com/dotnet/runtime/blob/969eaf85c5075dda9ea6241b61d42ba92f9dc9db/Directory.Build.props#L157

`'$(_runtimeOS)' == 'linux-musl' or $(_runtimeOS.StartsWith('alpine'))`

ayakael replied:
Copy that, thanks :)

ayakael commented:
Alpine now forces setting linker arg '--as-needed`, which removes the link forced by `-lucontext` in `mono.proj`. According to Alpine devs, this is an issue with the dotnet build system. Any pointers on how I should fix this? I've tried also setting `Wl,--no-as-needed` to neutralize but alas that doesn't work.

ayakael commented:
On `s390x`, `libcoreclr.so` is not explicitely linking to `libucontext.so` due to `--as-needed`, despite build of runtime being successful: `Error relocating artifacts/bin/mono/Linux.s390x.Release/libcoreclr.so: setcontext: symbol not found`. What am I forgetting?

uweigand commented:
> On `s390x`, `libcoreclr.so` is not explicitely linking to `libucontext.so` due to `--as-needed`, despite build of runtime being successful: `Error relocating artifacts/bin/mono/Linux.s390x.Release/libcoreclr.so: setcontext: symbol not found`. What am I forgetting?

Can you share the actual full linker command line in effect when building `libcoreclr.so`.  With `--as-needed`, the order of objects and libraries on the command line becomes important.

uweigand commented:
> On `s390x`, `libcoreclr.so` is not explicitely linking to `libucontext.so` due to `--as-needed`, despite build of runtime being successful: `Error relocating artifacts/bin/mono/Linux.s390x.Release/libcoreclr.so: setcontext: symbol not found`. What am I forgetting?

Can you share the actual full linker command line in effect when building `libcoreclr.so`?  With `--as-needed`, the order of objects and libraries on the command line becomes important.

am11 commented:
> full linker command line

`grep -R as-needed artifacts/obj` is your friend to figure it out.

ayakael commented:
Thanks everyone, here it is:
```
artifacts/obj/mono/Linux.s390x.Release/CMakeFiles/CMakeOutput.log: "/usr/bin/ld" -pie -z now -z relro --hash-style=gnu --build-id --eh-frame-hdr -m elf64_s390 -dynamic-linker /lib/ld-musl-s3
90x.so.1 -o cmTC_55628 /usr/lib/Scrt1.o /usr/lib/crti.o /usr/bin/../lib/gcc/s390x-alpine-linux-musl/12.2.1/crtbeginS.o -L/usr/bin/../lib/gcc/s390x-alpine-linux-musl/12.2.1 -L/usr/bin/../lib/
gcc/s390x-alpine-linux-musl/12.2.1/../../../../s390x-alpine-linux-musl/lib -L/lib -L/usr/lib --as-needed --build-id=sha1 -lucontext --no-as-needed CMakeFiles/cmTC_55628.dir/CMakeCCompilerABI
.c.o -lssp_nonshared -lgcc --as-needed -lgcc_s --no-as-needed -lc -lgcc --as-needed -lgcc_s --no-as-needed /usr/bin/../lib/gcc/s390x-alpine-linux-musl/12.2.1/crtendS.o /usr/lib/crtn.o
artifacts/obj/mono/Linux.s390x.Release/CMakeFiles/CMakeOutput.log:  ignore line: [clang-15: warning: -Wl --no-as-needed: 'linker' input unused [-Wunused-command-line-argument]]
```
Indeed, --no-as-needed is after `--as-needed`

ayakael commented:
Fixed the issue via setting `-Wl,--no-as-needed` before `-Wl,-lucontext`, with the added `target_link_libraries(mono-sgen PRIVATE ucontext)`. Thanks for the pointers :)

ayakael commented:
Found a proper fix where `-lucontext` is not needed anymore. In line with previous suggestions, this makes linking less hacky, and more specific. @am11 Could you do a final review? It'd be nice to merge this.

akoeplinger committed:
Merge branch 'main' into mono-musl-support

akoeplinger reviewed (approved):
LGTM thank you!

akoeplinger committed:
Support alpine as TARGET_SYSTEM_NAME

akoeplinger committed:
Update CMakeLists.txt

Sapana-Khemkar commented on src/native/libs/System.Native/pal_io.c:
```diff
@@ -1325,7 +1325,7 @@ int32_t SystemNative_CopyFile(intptr_t sourceFd, intptr_t destinationFd, int64_t
     if (sourceLength != 0)
     {
-        while ((ret = ioctl(outFd, FICLONE, inFd)) < 0 && errno == EINTR);
+        while ((ret = ioctl(outFd, (int)FICLONE, inFd)) < 0 && errno == EINTR);
```
@ayakael 
Why are you casting FICLONE to int? Second argument of ioctl function is unsigned long and not int. 
This caused ppc64le build error. Can you please revert this change?

ayakael replied:
This was a proposal by @am11, as build failed without setting `Wno-sign-conversion`. See: https://github.com/dotnet/runtime/pull/76500#discussion_r1010819744. 

Here is the exact error:
```
/var/build/dotnet7/testing/dotnet7-stage0/src/dotnet-e6dd91c290b808f971a1ac69c2fb29395bbf1051/src/runtime/src/native/libs/System.Native/pal_io.c(1270,36): error GBBD8A67C: implicit conversion changes signedness: 'unsigned long' to 'int' [-Wsign-conversion] [/var/build/dotnet7/testing/dotnet7-stage0/src/dotnet-e6dd91c290b808f971a1ac69c2fb29395bbf1051/src/runtime/src/native/libs/build-native.proj]
/var/build/dotnet7/testing/dotnet7-stage0/src/dotnet-e6dd91c290b808f971a1ac69c2fb29395bbf1051/src/runtime/src/native/libs/System.Native/pal_io.c(1270,36): error GBBD8A67C: implicit conversion changes signedness: 'unsigned long' to 'int' [-Wsign-conversion] [/var/build/dotnet7/testing/dotnet7-stage0/src/dotnet-e6dd91c290b808f971a1ac69c2fb29395bbf1051/src/runtime/src/native/libs/build-native.proj]
```

am11 replied:
If you change last argument to `unsigned long` here: https://github.com/dotnet/runtime/blob/cbc8695ae0c8c2c2d1ac1fc4546d81e0967ef716/src/native/libs/System.Native/pal_io.c#L76 without this `(int)FICLONE` cast, does it fix both `linux-ppc64le` and `linux-musl-ppc64le` builds?

uweigand replied:
That would be incorrect - the `_IOW` macro uses `sizeof` its last argument as a component of the resulting value, so changing that size from 4 to 8 will result in an incorrect value of `FICLONE`.

This issue seems to be caused by some difference in the definition of `ioctl` with musl.  In glibc, we have this:
```
extern int ioctl (int __fd, unsigned long int __request, ...);
```
Is this declared differently in musl?

am11 replied:
It is a ppc64le specific issue, where musl and glibc have different definitions. Rest of the platform matrix (including linux-musl-x64/arm64/arm etc..) is fine either way.

> That would be incorrect - the _IOW macro uses sizeof

I see, so the current definition is also incorrect; should be `size_t` instead of `int`, right?

uweigand replied:
> It is a ppc64le specific issue, where musl and glibc have different definitions. Rest of the platform matrix (including linux-musl-x64/arm64/arm etc..) is fine either way.

The only difference between ppc64le and other platforms is the numerical value of `FICLONE` here.  On ppc64le, this value is `0x80049409` while on other platforms it is `0x40049409` due to a different definition of the `_IOW` macro.  This means on other platforms the value is the same interpreted as `int` and `unsigned long`, so there's no compiler warning.  On ppc64le, the value is negative interpreted as `int` and positive interpreted as `unsigned long`, that's why we get the warning. The type mismatch itself is currently present on all platforms.

> > That would be incorrect - the _IOW macro uses sizeof
> 
> I see, so the current definition is also incorrect; should be `size_t` instead of `int`, right?

No.  We must use `sizeof (int)`, i.e. 4, here; this flows into the construction of that value above.

uweigand replied:
> This issue seems to be caused by some difference in the definition of `ioctl` with musl. In glibc, we have this:
> 
> ```
> extern int ioctl (int __fd, unsigned long int __request, ...);
> ```
> 
> Is this declared differently in musl?

Indeed this seems to be problem.  In the musl headers I see instead:
```
int ioctl (int, int, ...);
```
I guess as long as the two headers differ, there's probably no way to fix the source to work with both libcs, except for using some #ifdef ...

ayakael commented:
Thank you for the fix!

Would it be worthwhile to backport this and https://github.com/dotnet/runtime/pull/82173 to `release/7.0` and `release/6.0` ? Since we target `s390x` and `ppc64le` on Alpine Linux, I'd prefer to upstream what I can. I know @am11 objected to this on the grounds that supporting a new platform goes beyond bug fixes, but I'd like to appeal given that these changes, for the most part, only express themselves when building on Alpine Linux.

am11 commented:
@ayakael, could you tell a bit why the s390x support in 7.0 and 6.0 is important? Normally folks tend to use "next version" for new (previously unsupported) platforms / architectures, e.g. libunwind 1.6 added riscv64 support, and Alpine package enabled it for that version: https://github.com/alpinelinux/aports/commit/15c364170b77740814b625f77a59fa75ea2f0ebc. They did not backport to the previous version. It is easy to remember which version added support for which architecture/platform; it is also something we can announce and include in the release notes etc.

Same in the runtime repo that historically, we don't backport new platform support. It makes sense because it gives us some time to test and fix bugs / improve implementation. e.g. so far, there is no support for linux-musl-s390x in our cross infra (`eng/common/cross/build-rootfs.sh`, `toolchain.sh`, dotnet/dotnet-buildtools-prereqs-docker etc.), which is something we have kept in sync so far and considered as part of the port. Most people working on runtime do not use the source-build infra for development.

am11 commented:
@ayakael, could you tell a bit why the s390x support in 7.0 and 6.0 is important? Normally folks tend to use "next version" for new (previously unsupported) platforms / architectures, e.g. libunwind 1.6 added riscv64 support, and Alpine package enabled it for that version: https://github.com/alpinelinux/aports/commit/15c364170b77740814b625f77a59fa75ea2f0ebc. They did not backport to the previous version. This way, it is easy to remember which version added support for which architecture/platform; it is also something we can announce and include in the release notes etc.

Same in the runtime repo that historically, we don't backport new platform support. It makes sense because it gives us some time to test and fix bugs / improve implementation. e.g. so far, there is no support for linux-musl-s390x in our cross infra (`eng/common/cross/build-rootfs.sh`, `toolchain.sh`, dotnet/dotnet-buildtools-prereqs-docker etc.), which is something we have kept in sync so far and considered as part of the port. Most people working on runtime do not use the source-build infra for development.

ayakael commented:
> @ayakael, could you tell a bit why the s390x support in 7.0 and 6.0 is important? Normally folks tend to use "next version" for new (previously unsupported) platforms / architectures, e.g. libunwind 1.6 added riscv64 support, and Alpine package enabled it for that version: [alpinelinux/aports@15c3641](https://github.com/alpinelinux/aports/commit/15c364170b77740814b625f77a59fa75ea2f0ebc). They did not backport to the previous version. This way, it is easy to remember which version added support for which architecture/platform; it is also something we can announce and include in the release notes etc.
> 
> Same in the runtime repo that historically, we don't backport new platform support. It makes sense because it gives us some time to test and fix bugs / improve implementation. e.g. so far, there is no support for linux-musl-s390x in our cross infra (`eng/common/cross/build-rootfs.sh`, `toolchain.sh`, dotnet/dotnet-buildtools-prereqs-docker etc.), which is something we have kept in sync so far and considered as part of the port. Most people working on runtime do not use the source-build infra for development.

When it comes to which platforms we support for dotnet, we try to keep parity with Fedora. Thus, once s390x was supported on `fedora-s390x`, representing almost all of engineering work, porting to `alpine-s390x` was a matter of ironing out the musl quirks(mainly the stack issue). My reading of this PR is very just through the lense of a bug fix, as the changes aren't as extensive as a new platform support. Indeed, if the testing infrastructure was there to support it, `linux-musl-s390x` support should've been developped along-side `linux-s390x`. That said, I get your point. We already include these patches in the Alpine package for dotnet6 and dotnet7, so it doesn't change anything on Alpine's side, other than sharing the work with other musl-based platforms.  

Thanks for mentioning the cross infra needs, it reminds me that I need to complete the work on https://github.com/dotnet/arcade/pull/11608
